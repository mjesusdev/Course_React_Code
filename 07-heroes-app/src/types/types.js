export const types = {
    login: '[auth] login',
    logout: ''
}