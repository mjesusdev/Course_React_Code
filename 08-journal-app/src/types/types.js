export const types = {
    login: '[Auth] Login',
    logout: '[Auth] Logout',

    uiSetError: '[UI] Set Error',
    uiRemoveError: '[UI] Remove Error',
}