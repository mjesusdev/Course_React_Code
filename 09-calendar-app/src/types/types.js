export const types = {
    uiOpenModal: '[ui] Open modal',
    uiCloseModal: '[ui] Close modal',

    eventSetActive: '[event] Set Active',
    eventAddNew: '[event] Add new',

    eventClearActiveEvent: '[event] Clear Active event',
    eventUpdated: '[event] Event updated',
    eventDeleted: '[event] Event deleted'
}