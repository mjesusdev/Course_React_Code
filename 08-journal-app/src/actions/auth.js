import { types } from "../types/types";
import { firebase, googleAuthProvider } from '../firebase/firebase-config';

// Sign in with Email & Password
export const startLoginEmailPassword = ( email, password ) => {
    return ( dispatch ) => {
        setTimeout(() => {
            dispatch( login(123, 'Peter') );
        }, 3000)
    }
}

// Register User with Email & Password
export const startRegisterUser = ( email, password, name ) => {
    return ( dispatch ) => {

        firebase.auth().createUserWithEmailAndPassword( email, password)
            .then( async ({ user }) => {

                await user.updateProfile({ displayName: name });

                dispatch(
                    login( user.uid, user.displayName )
                )
            })

            .catch( error => {
                console.log(error);
            })
    }
}

// Sign in with Google Account
export const startGoogleLogin = () => {
    return ( dispatch ) => {
        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( ({ user }) => {
                dispatch(
                    login( user.uid, user.displayName )
                )
            });
    }
}

// Export
export const login = ( uid, displayName ) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
})