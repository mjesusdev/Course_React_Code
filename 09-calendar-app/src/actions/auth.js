import { fetchWithToken, fetchWitoutToken } from "../helpers/fetch"
import { types } from "../types/types";
import Swal from 'sweetalert2';

export const startLogin = ( email, password ) => {
    return async( dispatch ) => {
        const resp = await fetchWitoutToken( 'auth', { email, password }, 'POST' );
        const body = await resp.json();
        
        if ( body.ok ) {
            // Save Token in LocalStorage ✅
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login({
                uid: body.uid,
                name: body.name
            }) );
        } else {
            Swal.fire('Error', body.msg, 'error');
        }
    }
}

export const startRegister = ( email, password, name ) => {
    return async( dispatch ) => {
        const resp = await fetchWitoutToken('auth/new', { email, password, name }, 'POST' );
        const body = await resp.json();

        if ( body.ok ) {
            // Save Token in LocalStorage ✅
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( register({
                uid: body.uid,
                name: body.name
            }) );
        } else {
            Swal.fire('Error', body.msg, 'error');
        }
    }
}

export const startChecking = () => {
    return async ( dispatch ) => {
        const resp = await fetchWithToken( 'auth/renew' );
        const body = await resp.json();
        
        if ( body.ok ) {
            // Save Token in LocalStorage ✅
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );
        
            dispatch( register({
                uid: body.uid,
                name: body.name
            }) );
        } else {
            dispatch( checkingFinish() );
        }
    }
}

const checkingFinish = () => ({
    type: types.authCheckingFinish
})

const login = ( user ) => ({ 
    type: types.authLogin,
    payload: user
});

const register = ( user ) => ({
    type: types.authStartRegister,
    payload: user
})