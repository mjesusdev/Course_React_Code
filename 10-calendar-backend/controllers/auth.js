const { response } = require('express');
const bcrypt = require('bcryptjs');
const User = require('../models/User');
const { generateJWT } = require('../helpers/jwt');

// Function for create user and save SECURE in BD ✅
const createUser = async (req, res = response ) => {
    const { email, password } = req.body;
    
    try{
        let user = await User.findOne({ email });

        if ( user ) {
            return res.status(400).json({
                ok: false,
                msg: 'User exists with this email ❗'
            })
        }
        
        user = new User( req.body );
        
        // Encrypt password
        const salt = bcrypt.genSaltSync();
        user.password = bcrypt.hashSync( password, salt );
        
        await user.save();

        // Generate JWT
        const token = await generateJWT( user.id, user.name );
        
        res.status(201).json({
            ok: true,
            uid: user.id,
            name: user.name,
            token
        });
    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: 'Please speak with the administrator'
        })
    }
};

// Funcion for login user in app ✅
const loginUser = async (req, res = response) => {
    const { email, password } = req.body;
    
    try {
        const user = await User.findOne({ email });
        
        if ( !user ) {
            return res.status(400).json({
                ok: false,
                msg: 'The User not exists with this email ❗'
            });
        }

        // Confirm passwords 🔐
        const validPassword = bcrypt.compareSync( password, user.password );

        if ( !validPassword ) {
            return res.status(400).json({
                ok: false,
                msg: 'Incorrect Password ❗'
            });
        }

        // Generate our JWT | TOKEN
        const token = await generateJWT( user.id, user.name );

        res.json({
            ok: true,
            uid: user.id,
            name: user.name,
            token
        });

    } catch (error) {
        res.status(500).json({
            ok: false,
            msg: 'Speak with the Administrator, please!'
        })
    }
}

// Function for renew token | LOGIN ✅
const renewToken = async (req, res = response ) => {
    const { uid, name } = req;

    // Generate new JWT and return in this 👇
    const token = await generateJWT( uid, name );

    res.json({
        ok: true,
        uid,
        name,
        token,
    })
}

module.exports = {
    createUser,
    loginUser,
    renewToken,
};