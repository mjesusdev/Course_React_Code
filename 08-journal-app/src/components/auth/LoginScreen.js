import React from 'react';
import { useDispatch } from 'react-redux';

import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { startLoginEmailPassword, startGoogleLogin } from '../../actions/auth';

export const LoginScreen = () => {

    const dispatch = useDispatch();

    const [ formValues, handleInputChange ] = useForm({
        email: 'prueba@gmail.com',
        password: '12345'
    });

    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        dispatch( startLoginEmailPassword( email, password ) );
    }

    const handleGoogleLogin = () => {
        dispatch( startGoogleLogin() );
    }

    return (
        <>
           <h3 className="auth__title">Login</h3>

           <form onSubmit= { handleLogin }>
               <input 
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={ email }
                    onChange={ handleInputChange }
                    />

               <input 
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={ password }
                    onChange={ handleInputChange }
                />

                <button 
                    type="submit" 
                    className="btn btn-primary btn-block"
                >
                        Login
                </button>

                <hr />
                <div className="auth__social-networks">
                    <p>Login with social networks</p>
                    <div 
                        className="google-btn"
                        onClick={ handleGoogleLogin }
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" 
                                alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link 
                    to="/auth/register"
                    className="link"
                >
                    Create new account
                </Link>
           </form>
        </>
    )
}
