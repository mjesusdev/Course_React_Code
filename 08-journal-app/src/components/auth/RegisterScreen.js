import React from 'react';
import { Link } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import validator from 'validator';
import Swal from 'sweetalert2';

import { setError, removeError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';
import { startRegisterUser } from '../../actions/auth';

export const RegisterScreen = () => {

    const dispatch = useDispatch();
    /* const { msgError } = useSelector( state => state.ui ); */

    const [ formValues, handleInputChange ] = useForm({
        name: 'Prueba',
        email: 'prueba@gmail.com',
        password: '123456',
        password2: '123456'
    });

    // Get Values
    const { name, email, password, password2 } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        
        if ( isFormValid() ) {
            dispatch(startRegisterUser( email, password, name ) );
        }
    }

    const isFormValid = () => {
        if ( name.trim().length === 0 ) {
            dispatch(setError('Name is required'));
            Swal.fire('Error', 'Name is required', 'error');
            return false;
        } else if ( !validator.isEmail(email) ) {
            dispatch(setError('Email is not valid'));
            Swal.fire('Error', 'Email is not valid', 'error');
            return false;
        } else if ( password !== password2 || password.length < 5) {
            dispatch(setError('Password should be at least 6 characters and match each other'));
            Swal.fire('Error', 'Password should be at least 6 characters and match each other', 'error');
            return false;
        }

        dispatch(removeError());
        return true;
    }

    return (
        <>
            <h3 className="auth__title">Register</h3>

            <form onSubmit={ handleRegister }>

                {/* {
                    msgError &&
                    (
                        <div className="auth__alert-error">
                            { msgError }
                        </div>
                    )
                } */}

                <input 
                    type="text"
                    placeholder="Name"
                    name="name"
                    className="auth__input"
                    autoComplete="off"
                    value={ name }
                    onChange={ handleInputChange }
                />

                <input 
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={ email }
                    onChange={ handleInputChange }
                />

                <input 
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={ password }
                    onChange={ handleInputChange }
                />

                <input 
                    type="password"
                    placeholder="Confirm password"
                    name="password2"
                    className="auth__input"
                    value={ password2 }
                    onChange={ handleInputChange }
                />

                <button 
                    type="submit" 
                    className="btn btn-primary btn-block mb-5"
                >
                    Register
                </button>

                <hr />

                <Link 
                    to="/auth/login"
                    className="link mt-5"
                >
                    Already registered?
                </Link>
            </form>
        </>
    )
}
