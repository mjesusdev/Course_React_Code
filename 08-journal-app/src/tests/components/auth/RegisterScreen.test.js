import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';

import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';

const middlewars = [thunk];
const mockStore = configureStore(middlewars);

const initState = {
    auth : {},
    ui: {
        loading: false,
        msgError: null
    }
};

let store = mockStore(initState);

describe('Test in <RegisterScreen />', () => {
    
    const wrapper = mount(
        <Provider store = { store }>
            <MemoryRouter>
                <RegisterScreen />
            </MemoryRouter>
        </Provider>
    )
    
    test('should show correctly', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('should do dispatch of the action', () => {
        const emailField = wrapper.find('input[name="email"]');
        emailField.simulate('change', {
            target: {
                value: '',
                name: 'email'
            }
        });

        wrapper.find('form').simulate('submit', {
            preventDefault() {}
        });

        const actions = store.getActions();

        expect( actions[0] ).toEqual({
            type: types.uiSetError,
            payload: 'Email is not valid'
        });
    });

    /*     test('should show alert box with the error', () => {
        const initState = {
            auth : {},
            ui: {
                loading: false,
                msgError: 'Email is not correct'
            }
        };
        
        let store = mockStore(initState);

        const wrapper = mount(
            <Provider store = { store }>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );

        expect( wrapper.find('.auth__alert-error').exists() ).toBe(true);
        expect( wrapper.find('.auth__alert-error').text().trim() ).toBe( initState.ui.msgError );
    }); */
});