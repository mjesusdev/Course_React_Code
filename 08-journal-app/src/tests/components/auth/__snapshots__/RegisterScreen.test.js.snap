// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <RegisterScreen /> should show correctly 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn animate_faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Prueba"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="prueba@gmail.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <hr />
    <a
      className="link mt-5"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
